lib/wasi/src/state/mod.rs

```diff
@@ -819,6 +819,56 @@ impl WasiFs {
                                     relative_path: link_value,
                                 }
                             } else {
+                                #[cfg(unix)]
+                                {
+                                    use std::os::unix::fs::FileTypeExt;
+                                    let file_type: __wasi_filetype_t = if file_type.is_char_device()
+                                    {
+                                        __WASI_FILETYPE_CHARACTER_DEVICE
+                                    } else if file_type.is_block_device() {
+                                        __WASI_FILETYPE_BLOCK_DEVICE
+                                    } else if file_type.is_fifo() {
+                                        // FIFO doesn't seem to fit any other type, so unknown
+                                        __WASI_FILETYPE_UNKNOWN
+                                    } else if file_type.is_socket() {
+                                        // TODO: how do we know if it's a `__WASI_FILETYPE_SOCKET_STREAM` or
+                                        // a `__WASI_FILETYPE_SOCKET_DGRAM`?
+                                        __WASI_FILETYPE_SOCKET_STREAM
+                                    } else {
+                                        unimplemented!("state::get_inode_at_path unknown file type: not file, directory, symlink, char device, block device, fifo, or socket");
+                                    };
+
+                                    let kind = Kind::File {
+                                        handle: None,
+                                        path: file.clone(),
+                                        fd: None,
+                                    };
+                                    let new_inode = self.create_inode_with_stat(
+                                        kind,
+                                        false,
+                                        file.to_string_lossy().to_string(),
+                                        __wasi_filestat_t {
+                                            st_filetype: file_type,
+                                            ..__wasi_filestat_t::default()
+                                        },
+                                    );
+                                    if let Kind::Dir {
+                                        ref mut entries, ..
+                                    } = &mut self.inodes[cur_inode].kind
+                                    {
+                                        entries.insert(
+                                            component.as_os_str().to_string_lossy().to_string(),
+                                            new_inode,
+                                        );
+                                    } else {
+                                        unreachable!(
```

Comment:
Just a comment, I'm not sure whether this is reachable or not? If the place in the wasi-fs where the host path is being injected is controlled by the user then it seems like it should be a full Err? Otherwise, if this is part of mirroring the host paths then this is fine as is.

Reply:
It shouldn't be reachable without unsafe I believe: it's not possible to put files in the virtual root right now, they have to come from a preopened directory and that directory will be mounted at `/`.